Enable S32K3 WKPU IP block to pass interrupt trigger through GPIO

manuargue opened the pull request:
This PR introduces an interrupt controller driver for the NXP S32 WKPU and integrates it with the GPIO driver to extend the number of GPIO pins that can interrupt the core.

GPIO pins that are input-interrupt capable support triggering SIUL2 EIRQ interrupts (existing interrupt controller implementation) and/or WKPU interrupts (introduced in this patch). To configure WKPU to trigger an interrupt from a GPIO pin, a SoC-specific flag must be supplied to the `gpios` specifier to instruct the GPIO driver which interrupt controller to use.

> [!IMPORTANT]
> WKPU supports triggering interrupts from certain input pads that support this function, as well as generating wake-up events to the power management domain. This patch only adds WKPU functionality as an interrupt controller to extend on this SoC the number of input pads that can interrupt the core. Power management functionalities (ie. wake-up events to power management domain) are not supported and are not in the scope of this PR.

Tests:
```
> west twister -p mr_canhubk3 -T .\tests\drivers\gpio\gpio_basic_api --device-testing --device-serial=COM4 --fixture=gpio_loopback
INFO    - Using Ninja..
INFO    - Zephyr version: v3.5.0-rc1-95-g066503e2555a
INFO    - Using 'zephyr' toolchain.
...
INFO    - Total complete:    3/   3  100%  skipped:    1, failed:    0, error:    0
INFO    - 3 test scenarios (3 test instances) selected, 1 configurations skipped (1 by static filter, 0 at runtime).
INFO    - 2 of 3 test configurations passed (100.00%), 0 failed, 0 errored, 1 skipped with 0 warnings in 291.60 seconds
INFO    - In total 10 test cases were executed, 5 skipped on 1 out of total 626 platforms (0.16%)
INFO    - 2 test configurations executed on platforms, 0 test configurations were only built.
```

manuargue-nxp committed:
intc: add NXP S32 WKPU interrupt controller driver

Introduce an interrupt controller for the NXP S32 WKPU peripheral
that can be integrated with GPIO to trigger interrupts through
external interrupt pad inputs.

WKPU can trigger interrupts from certain input pads that support this
function, as well as wake-up events to the power management domain. This
patch only adds WKPU functionality as an interrupt controller to extend
the number of input pads that can interrupt the core. Power management
functionalities are not supported.

Signed-off-by: Manuel Argüelles <manuel.arguelles@nxp.com>

manuargue-nxp committed:
gpio: nxp_s32: prepare to support multiple interrupt controllers

Refactor external interrupts infrastructure to prepare supporting
multiple interrupt controllers.

Signed-off-by: Manuel Argüelles <manuel.arguelles@nxp.com>

manuargue-nxp committed:
gpio: nxp_s32: support passing external interrupts to WKPU

Extend the NXP S32 GPIO driver to be able to route external interrupts
to either SIUL2 EIRQ interrupt controller or, when available on the
SoC, WKPU interrupt controller.

Since WKPU can support up to 64 external interrupt sources and SIUL2
EIRQ up to 32, gpio_get_pending_int() is removed and the interrupt
controller specific API must be used instead.

Signed-off-by: Manuel Argüelles <manuel.arguelles@nxp.com>

manuargue-nxp committed:
nxp_s32k344: add external interrupts for WKPU

Define WKPU interrupt controller node and its respective interrupt
sources mapping to GPIO pins.

Signed-off-by: Manuel Argüelles <manuel.arguelles@nxp.com>

manuargue-nxp committed:
mr_canhubk3: document WKPU interrupt controller support

Enhance the documentation by providing an explanation of WKPU
interrupt controller support and how it integrates with GPIO.

Signed-off-by: Manuel Argüelles <manuel.arguelles@nxp.com>

manuargue-nxp committed:
tests: gpio_basic_api: add WKPU interrupts test for mr_canhubk3

Use the same pins as when testing with SIUL2 EIRQ interrupt
controller, but instead test routing the external interrupts
to WKPU controller.

Signed-off-by: Manuel Argüelles <manuel.arguelles@nxp.com>

manuargue-nxp committed:
tests: gpio_basic_api: use GPIO flags from DT

When configuring in/out pins for callback tests, pass the GPIO flags
coming from the in/out-gpios respective nodes.

Signed-off-by: Manuel Argüelles <manuel.arguelles@nxp.com>

gmarull reviewed (changes requested):
what about `wakeup-source` standard property?

manuargue commented:
> what about `wakeup-source` standard property?

I have added the missing property to allow declaring the GPIO port wake-up capable and updated the test to use it.

gmarull commented:
In general I think we need a solution that is cross-vendor compatible. We have now STM32 using HAL directly in their power off sample, Nordic using gpio_interrupt_configure (it works by chance, because what is configured turns out to allow wake-up from pins) and now NXP introducing a vendor specific option.

Please, let's propose something that can be made portable. There are enough SoCs with a similar feature.

bperseghetti commented:
@manuargue Thanks for getting this much needed PR in! This is a necessity to allow for using the shields on the CANHUBK3 boards!

> In general I think we need a solution that is cross-vendor compatible. We have now STM32 using HAL directly in their power off sample, Nordic using gpio_interrupt_configure (it works by chance, because what is configured turns out to allow wake-up from pins) and now NXP introducing a vendor specific option.
> 
> Please, let's propose something that can be made portable. There are enough SoCs with a similar feature.

@gmarull were you looking at bringing an RFC forward for doing that? I assume your suggestion to find a better eventual combined solution is non-blocking for this PR.

gmarull commented:
> @manuargue Thanks for getting this much needed PR in! This is a necessity to allow for using the shields on the CANHUBK3 boards!
> 
> > In general I think we need a solution that is cross-vendor compatible. We have now STM32 using HAL directly in their power off sample, Nordic using gpio_interrupt_configure (it works by chance, because what is configured turns out to allow wake-up from pins) and now NXP introducing a vendor specific option.
> > Please, let's propose something that can be made portable. There are enough SoCs with a similar feature.
> 
> @gmarull were you looking at bringing an RFC forward for doing that? I assume your suggestion to find a better eventual combined solution is non-blocking for this PR.

No, it is a blocking request. I think we have enough vendor-specific stuff to finally find a common ground.

PetervdPerk-NXP commented:
> No, it is a blocking request. I think we have enough vendor-specific stuff to finally find a common ground.

You've reviewed this driver and saw it's just a pin interrupt source coming from the WKPU IP unit? 
This PR doesn't implement a wake-up functionality at all it's just forward the interrupts from WKPU block acting like an external interrupt (EIRQ).

> Nordic using gpio_interrupt_configure (it works by chance, because what is configured turns out to allow wake-up from pins) and now NXP introducing a vendor specific option. 

We're simply not implementing a wake-up functionality so I don't see a common ground you're referring too also the S32K344 doesn't support a low-power / sleep mode in Zephyr at the moment so it wouldn't be applicable.

manuargue commented:
> > No, it is a blocking request. I think we have enough vendor-specific stuff to finally find a common ground.
> 
> You've reviewed this driver and saw it's just a pin interrupt source coming from the WKPU IP unit? This PR doesn't implement a wake-up functionality at all it's just forward the interrupts from WKPU block acting like an external interrupt (EIRQ).

In addition to comments above, I've have noted this in the board's doc already to make it clear https://github.com/zephyrproject-rtos/zephyr/commit/c32c34b50c8b17b573924d77677d5e3e3bcb796f#diff-98d098f031b3ee9096999b5e0c03263a156546694f48e1feeef4f72a30cb06f0R79

gmarull commented:
> > No, it is a blocking request. I think we have enough vendor-specific stuff to finally find a common ground.
> 
> You've reviewed this driver and saw it's just a pin interrupt source coming from the WKPU IP unit? This PR doesn't implement a wake-up functionality at all it's just forward the interrupts from WKPU block acting like an external interrupt (EIRQ).
> 

from your binding:

> GPIOs that are supported as external wake-up sources

PetervdPerk-NXP commented:
So that's simply wrong wording from the author. 
The WKPU IP Block can do that but this driver only does the interrupt part because on the S32K344 not all pins can be used as an interrupt only pins that can be muxed to either EIRQ and WKPU block can generate an interrupt.

You okay with this PR if we use different wording and emphasize only that the driver implements the interrupt part of the WKPU mux mode?

gmarull commented:
> So that's simply wrong wording from the author. The WKPU IP Block can do that but this driver only does the interrupt part because on the S32K344 not all pins can be used as an interrupt only pins that can be muxed to either EIRQ and WKPU block can generate an interrupt.
> 
> You okay with this PR if we use different wording and emphasize only that the driver implements the interrupt part of the WKPU mux mode?

So, what does the vendor specific GPIO flag do? Isn't it used to configure the GPIO as a wakeup source?

PetervdPerk-NXP commented:
> So, what does the vendor specific GPIO flag do? Isn't it used to configure the GPIO as a wakeup source?
Which specific flag are you referring too?

gmarull commented:
> > So, what does the vendor specific GPIO flag do? Isn't it used to configure the GPIO as a wakeup source?
> 
> Which specific flag are you referring too?

`NXP_S32_GPIO_INT_WKPU`

also...

> Some GPIO pins support to generate external interrupts and/or wake-up interrupts. To configure a GPIO to trigger wake-up interrupts, follow the instructions in :zephyr_file:`dts/bindings/gpio/nxp,s32-gpio.yaml`.

So **this is** about configuring a GPIO as a wake-up source, something many vendors do. I'm asking to propose a common ground to create a portable solution. Unless I'm totally wrong, which can happen, my -1 stands.

PetervdPerk-NXP commented:
```
/** Interrupt routed to the SIUL2 External Interrupt controller */
#define NXP_S32_GPIO_INT_EIRQ	(0x0U << NXP_S32_GPIO_INT_TYPE_POS)
/** Interrupt routed to the WKPU controller */
#define NXP_S32_GPIO_INT_WKPU	(0x1U << NXP_S32_GPIO_INT_TYPE_POS)
```
So that's to specify the interrupt controller for a interrupt on a specific pin. That's because some pins support both WKPU & EIRQ, some only EIRQ, some only WKPU, some neither. For example:

| Pin   | EIRQ | WKPU | IRQ possible |
|-------|------|------|--------------|
| PTA5  | Yes  | No   | Yes          |
| PTC18 | No   | Yes  | Yes          |
| PTA30 | Yes  | Yes  | Yes          |
| PTC30 | No   | No   | No           |

> Some GPIO pins support to generate external interrupts and/or wake-up interrupts. To configure a GPIO to trigger wake-up interrupts, follow the instructions in :zephyr_file:dts/bindings/gpio/nxp,s32-gpio.yaml.

So the `nxp,s32-gpio.yaml` doesn't describe this so this is faulty, hence I'm proposing we take out the wake-up part and only focus on interrupts.

manuargue commented:
GPIOs can raise interrupts through two different hw blocks, the external interrupt controller and/or the wakeup controller. `NXP_S32_GPIO_INT_WKPU`  tells the GPIO driver that the interrupt to be generated is through the wakeup controller and not through the external interrupt controller. GPIOs that are wake-up source capable (terminology from S32K3 RM) are able generate **interrupts** or wakeup **events**. This PR is about generating interrupts, not wakeup events. I can change the wording in `nxp,s32-gpio.yaml` from `To configure a GPIO to trigger wake-up interrupts..` to `To configure a GPIO to trigger an interrupt through the wakeup controller` if semantics is the concern here.

dleach02 commented:
@gmarull, please revisit this PR. We need to get it merged before the merge freeze tomorrow.

gmarull commented:
> @gmarull, please revisit this PR. We need to get it merged before the merge freeze tomorrow.

I understand your rush to merge things, but this feature is still unclear to me, so please, make at least the wording clear so I can review it again. I’d suggest to maintain a synced fork with your upmergeable patches to decouple your internal timelines from Zephyr.

bperseghetti commented:
> > @gmarull, please revisit this PR. We need to get it merged before the merge freeze tomorrow.
> 
> I understand your rush to merge things, but this feature is still unclear to me, so please, make at least the wording clear so I can review it again. I’d suggest to maintain a synced fork with your upmergeable patches to decouple your internal timelines from Zephyr.

@gmarull What aspect of this feature is unclear to you that we can help clarify so you understand it better, did you see the further explanation from Peter and [Manuel](https://github.com/zephyrproject-rtos/zephyr/pull/62623#issuecomment-1737119056). This is not a rush and the point of a merge window before the release freeze is to merge meaningful things, this is a needed meaningful thing.

I definitely understand the desire for better unification across vendors, but for our understanding please show us an exact example of where using a specific IP block like WKPU for passing through a standard non-wakeup GPIO interrupt is replicable/generalizable across multiple vendors. If you cannot show how this could be meaningfully generalized with a direct example or if maybe the concept of what we need/are doing is not what you initially thought/understood (the passthrough of the WKPU IP blocks GPIo interrupt function that we are describing) I would ask that you consider removing the block as there would be no logical basis for keeping it otherwise.

manuargue commented:
Pushed the following updates with the intention to clarify the ambiguities around the interrupt controller being introduced in this PR:

- improved wording and remove the use of "wakeup" terminology from bindings, docs, commit msgs, etc. to make clear the intention of this patch is to provide only a new interrupt controller integrated with GPIO.
- rename `nxp,s32-gpio.yaml` introduced properties to `nxp,wkpu` / `nxp,wkpu-interrupts` and remove any "wakeup" reference.

gmarull reviewed (commented):
So I still have one basic question. Why do you want to use WKPU (Wakeup Unit) with GPIOs, if not to configure them as wake-up sources?

![image](https://github.com/zephyrproject-rtos/zephyr/assets/25011557/7fa1cad0-91f1-41ce-86e1-3985bd0438cf)

PetervdPerk-NXP commented:
> So I still have one basic question. Why do you want to use WKPU (Wakeup Unit) with GPIOs, if not to configure them as wake-up sources?

Few reasons.

On the S32K344 every pin is by definition an GPIO due to how SIUL2 is designed.
![image](https://github.com/zephyrproject-rtos/zephyr/assets/57130844/4df6454b-5f32-4a62-8079-63debba05f42)


As described here https://github.com/zephyrproject-rtos/zephyr/pull/62623#issuecomment-1737111862 not all pins on the S32K344 can generate an interrupt for example pin PTC18 can only interrupt but has to go through WKPU interrupt controller.
![image](https://github.com/zephyrproject-rtos/zephyr/assets/57130844/3b374aa7-3b4d-4eb0-b9ee-8a3d9c5f95be)


For now there's no intention to implement low-power modes for the S32K344 therefore there's no need for wake-up sources.

gmarull commented:
So if some pins do require WKPU to generate interrupts, and these are known, why don't you handle them within the driver, using regular IRQ attributes? I mean, if the driver gets asked to configure interrupts for PTC18, just use the WKPU route.

PetervdPerk-NXP commented:
> So if some pins do require WKPU to generate interrupts, and these are known, why don't you handle them within the driver, using regular IRQ attributes? I mean, if the driver gets asked to configure interrupts for PTC18, just use the WKPU route.

So some pins can do both a WKPU and EIRQ interrupt unfortunately, thus the idea was to make it explicit in DTS to keep it simple. Otherwise you'll get a complex initialization scheme with big lookup tables to figure out the routing.

Furthermore this PR brings a dedicated WKPU ip block driver in, if the wake-up rewrite you were proposing comes in and on a later moment we want to implement low-power mode for S32K344 we would to have to split out anyhow.

gmarull commented:
> > So if some pins do require WKPU to generate interrupts, and these are known, why don't you handle them within the driver, using regular IRQ attributes? I mean, if the driver gets asked to configure interrupts for PTC18, just use the WKPU route.
> 
> So some pins can do both a WKPU and EIRQ interrupt unfortunately, thus the idea was to make it explicit in DTS to keep it simple. Otherwise you'll get a complex initialization scheme with big lookup tables to figure out the routing.
> 

So why one would use WKPU over EIRQ, if EIRQ is available?

PetervdPerk-NXP commented:
> So why one would use WKPU over EIRQ, if EIRQ is available?

If someone wants to use wake-up in the future?

It's a matter of preference I guess, I would see WKPU and EIRQ and 2 independent interrupt controllers.
Also this fits nicely the device tree philosophy zephyr is following.

Merging both into a single driver would hide the logic and make it ambiguous.

bperseghetti commented:
> > So why one would use WKPU over EIRQ, if EIRQ is available?
> 
> If someone wants to use wake-up in the future?
> 
> It's a matter of preference I guess, I would see WKPU and EIRQ as 2 independent interrupt controllers.
> Also this fits nicely the device tree philosophy zephyr is following.
> 
> Merging both into a single driver would hide the logic and make it ambiguous.

@gmarull it appears that the technical aspects of this have all been explained in detail along with the need for it. Do you have any more questions and/or are you able to remove your block now that it has been better explained/presented?

gmarull commented:
> > > So why one would use WKPU over EIRQ, if EIRQ is available?
> > 
> > 
> > If someone wants to use wake-up in the future?
> > It's a matter of preference I guess, I would see WKPU and EIRQ as 2 independent interrupt controllers.
> > Also this fits nicely the device tree philosophy zephyr is following.
> > Merging both into a single driver would hide the logic and make it ambiguous.
> 
> @gmarull it appears that the technical aspects of this have all been explained in detail along with the need for it. Do you have any more questions and/or are you able to remove your block now that it has been better explained/presented?

No. Please, read my answers above.

manuargue commented:
> This is precisely what I'm asking NOT to do. Zephyr needs a proper wake-up generalization, not yet another vendor-specific solution to the same problem.

@gmarull this PR is not providing a solution for power management wake-up functionalities. The intention is to extend the existing support of GPIO input interrupts as it is a necessity to bring up boards and shields using the S32K3 SoCs.

gmarull commented:
> > This is precisely what I'm asking NOT to do. Zephyr needs a proper wake-up generalization, not yet another vendor-specific solution to the same problem.
> 
> @gmarull this PR is not providing a solution for power management wake-up functionalities. The intention is to extend the existing support of GPIO input interrupts as it is a necessity to bring up boards and shields using the S32K3 SoCs ([#62623 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/62623#issuecomment-1719943612)).

So why is a new vendor-specific flag required? Can't driver decide internally what to choose? E.g. configure WKPU for pins that only have this route, EIRQ for the rest, all using standard interrupt GPIO flags. "User can decide" choice seems a weak argument because from what I can read, WKPU is a "Wakeup Unit".

PetervdPerk-NXP commented:
> So why is a new vendor-specific flag required? Can't driver decide internally what to choose? 

Because this seems to be S32K3XX vendor specific thing also it nicely ends up in the DTS 
Also why add more logic to the driver, "less is more" would be my approach.

> Zephyr needs a proper wake-up generalization, not yet another vendor-specific solution to the same problem.

That's a problem you've indentified but over again we've said we're not implementing wake-up. But are you now blocking this PR to get someone to do a PM wake-up generalization rewrite? Isn't that up to the PM subsys maintainers?

gmarull reviewed (dismissed):
the current implementation opens up the possibility to configure GPIOs as wake-up sources using a vendor-specific flag, even if _unwanted_.

# -1 stands

PetervdPerk-NXP commented:
> the current implementation opens up the possibility to configure GPIOs as wake-up sources using a vendor-specific flag, even if unwanted.

Many things are possible, but with the current PR the scenario you describe isn't possible.
So I'm not sure were you getting at.

mmahadevan108 commented on tests/drivers/gpio/gpio_basic_api/src/test_gpio.h:
```diff
@@ -25,7 +25,9 @@
 #define DEV_IN DT_GPIO_CTLR(DT_INST(0, test_gpio_basic_api), in_gpios)
 #define DEV DEV_OUT /* DEV_OUT should equal DEV_IN, we test for this */
 #define PIN_OUT DT_GPIO_PIN(DT_INST(0, test_gpio_basic_api), out_gpios)
+#define PIN_OUT_FLAGS DT_GPIO_FLAGS(DT_INST(0, test_gpio_basic_api), out_gpios)
```
Why is this needed?

mmahadevan108 commented on tests/drivers/gpio/gpio_basic_api/src/test_callback_trigger.c:
```diff
@@ -54,7 +54,7 @@ static int test_callback(int mode)
 
 	/* 2. configure PIN_IN callback and trigger condition */
-	rc = gpio_pin_configure(dev, PIN_IN, GPIO_INPUT);
+	rc = gpio_pin_configure(dev, PIN_IN, (GPIO_INPUT | PIN_IN_FLAGS));
```
Has this test been looked at on platforms that do not have this flag?

manuargue replied:
`PIN_IN_FLAGS` will be the GPIO flags of `in-gpios` node or zero if none, pls see https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/devicetree/gpio.h#L227

mmahadevan108 commented on tests/drivers/gpio/gpio_basic_api/testcase.yaml:
```diff
@@ -16,3 +16,7 @@ tests:
     platform_allow: nrf52840dk_nrf52840 nrf52_bsim
     extra_args: "DTC_OVERLAY_FILE=boards/nrf52840dk_nrf52840.overlay;\
                  boards/nrf52840dk_nrf52840_sense_edge.overlay"
+
+  drivers.gpio.mr_canhubk3_wkpu:
+    platform_allow: mr_canhubk3
+    extra_args: "DTC_OVERLAY_FILE=boards/mr_canhubk3_wkpu.overlay"
```
Why is this needed. Can we delete this?

manuargue replied:
this is adding a new test scenario for the `mr_canhubk3` board to test with WKPU interrupts. The default test scenario is to test with EIRQ interrupts:
https://github.com/zephyrproject-rtos/zephyr/blob/main/tests/drivers/gpio/gpio_basic_api/boards/mr_canhubk3.overlay#L16-L21

MaureenHelm reviewed (dismissed):
> Zephyr needs a proper wake-up generalization, not yet another vendor-specific solution to the same problem.

I don't think anyone disagrees on the need to generalize wakeup sources, but that doesn't need to be solved here. All this PR does is enable another interrupt controller for GPIO interrupts on the S32. Although some pins on the SoC can be routed to either interrupt controller, the pins that @bperseghetti wants to use can only be routed to the WKPU interrupt controller.

dleach02 commented:
@gmarull, thank you for your review and feedback. The conversation thread on this PR took, what I believed, to be some unfortunate directions with the limited communication clarity GitHub PR's provides. In the future, we will endeavor to raise these conversations to the dev-review meeting more quickly to have a wider bandwidth discussion. I am disappointed that a consensus could not be reached on this PR and I acknowledge that you have reiterated your -1 but as the maintainer of this subsystem, I'm exercising my right to remove your block.

With respect to your desire to have some common recipe for power management wakeup support, I do not disagree. We will be introducing a separate PR to introduce some GPIO wakeup control for our RT595 platform we did for the ZDS demo and will welcome your feedback there if you so choose to review it.

gmarull reviewed (dismissed):
sorry, but please do not dismiss reviews like this. Escalate this properly at least.

gmarull commented:
> > Zephyr needs a proper wake-up generalization, not yet another vendor-specific solution to the same problem.
> 
> I don't think anyone disagrees on the need to generalize wakeup sources, but that doesn't need to be solved here. All this PR does is enable another interrupt controller for GPIO interrupts on the S32. Although some pins on the SoC can be routed to either interrupt controller, the pins that @bperseghetti wants to use can only be routed to the WKPU interrupt controller.

This has other solutions that don't open the door to use the vendor flag to configure GPIOs as a wake-up source.

gmarull commented:
> @gmarull, thank you for your review and feedback. The conversation thread on this PR took, what I believed, to be some unfortunate directions with the limited communication clarity GitHub PR's provides. In the future, we will endeavor to raise these conversations to the dev-review meeting more quickly to have a wider bandwidth discussion. I am disappointed that a consensus could not be reached on this PR and I acknowledge that you have reiterated your -1 but as the maintainer of this subsystem, I'm exercising my right to remove your block.
> 
> You have the option to escalate your -1 to the TSC if you desire. The PR already has the TSC attention tag but please comment in this PR or the TSC channel if you wish to exercise your option.
> 
> With respect to your desire to have some common recipe for power management wakeup support, I do not disagree. We will be introducing a separate PR to introduce some GPIO wakeup control for our RT595 platform we did for the ZDS demo and will welcome your feedback there if you so choose to review it.

AFAIK, you're not the maintainer of GPIO nor interrupt controllers. Am I missing something? Also, where are the meeting notes that led to this "decision"? So, do vendors have the privilege now to dismiss at their will to get things in quick? Is this a new Zephyr rule?

fabiobaltieri commented:
> This has other solutions that don't open the door to use the vendor flag to configure GPIOs as a wake-up source.

Hey, trying to follow the conversation, could you be more explicit about the suggested path forward? I don't find that super clear, that's probably nested in the comments.

Are you asking to use the `wakeup-source` flag instead of the custom GPIO bits?

fabiobaltieri commented:
Also about the dev-review meeting, it only helps so much to do them without al the parties involved, for cases like that it'd be a good idea to ask for explicit acknowledge from the reviewers rather than assuming everyone saw the dev-review tag. I'd suggest a one-off separate talk before the TSC if we can't unstuck the situation here, happy to help organize in that case but maybe first let's see if we can get this to move.

gmarull commented:
> > This has other solutions that don't open the door to use the vendor flag to configure GPIOs as a wake-up source.
> 
> Hey, trying to follow the conversation, could you be more explicit about the suggested path forward? I don't find that super clear, that's probably nested in the comments.
> 
> Are you asking to use the `wakeup-source` flag instead of the custom GPIO bits?

So from what I understood in this PR, some pins like PCT18 can only generate an external interrupt via WKPU unit, not EIRQ. WKPU is the "Wakeup Unit", which, according to the RM, it seems to be a unit responsible for configuring multiple wake-up sources. This PR _resolves_ this problem by letting the user specify which _unit_ the users want. However, this has a problem: it opens the door to configure any GPIO as a wake-up source using vendor-specific flags. As I've mentioned, I'm against opening the door to yet another flavor to configure GPIOs as wake-up sources because let's be honest, if this is available, why will a vendor care about finding a general solution that will take time, effort, cross-vendor coordination, etc.?

Since the "limited" pins like PCT18 are in theory known, I've suggested that the driver deals with this internally. If it is requested to configure the interrupts for one of such pins, simply use WKPU, otherwise EIRQ. This makes interrupt configuration transparent to the user, no need for new DT flags, etc. I haven't heard why this option is not viable, for example, so I won't let my -1 to be dismissed just because some vendors have the rush to release.

Please correct me if I got something wrong. I haven't read the whole RM, just a few pages and this PR.

PetervdPerk-NXP commented:
> WKPU is the "Wakeup Unit", which, according to the RM, it seems to be a unit responsible for configuring multiple wake-up sources. 

Can you stop being so naive, and read further than the word wake-up. Indeed the functionality of the WKPU "Wakeup unit" is to provide wake-up events **and interrupts**. Which has been repeatedly commented here above.

This driver only **implements interrupts, no wake-up functionality.**

> However, this has a problem: it opens the door to configure any GPIO as a wake-up source using vendor-specific flags. 

Just block any PR that intends to do that, but reasoning that this is a possibility now is just invalid.

> Since the "limited" pins like PCT18 are in theory known, I've suggested that the driver deals with this internally. If it is requested to configure the interrupts for one of such pins, simply use WKPU, otherwise EIRQ. This makes interrupt configuration transparent to the user, no need for new DT flags, etc.

We've dismissed that suggestion from you before. Since it would be hiding the logic and makes it not transparent at all. Thus we make it explicit that the dts has to specifiy which interrupt controller has to be used for a certain pin.

fabiobaltieri commented:
> > Since the "limited" pins like PCT18 are in theory known, I've suggested that the driver deals with this internally. If it is requested to configure the interrupts for one of such pins, simply use WKPU, otherwise EIRQ. This makes interrupt configuration transparent to the user, no need for new DT flags, etc.
> 
> We've dismissed that suggestion from you before. Since it would be hiding the logic and makes it not transparent at all. Thus we make it explicit that the dts has to specifiy which interrupt controller has to be used for a certain pin.

Coud you summarize/expand on that front? Sorry joining late to the party and building context. At first approximation hiding that logic and make the switch happen automatically does not sound too bad, what else is in the picture?

PetervdPerk-NXP commented:
> Coud you summarize/expand on that front? Sorry joining late to the party and building context. At first approximation hiding that logic and make the switch happen automatically does not sound too bad, what else is in the picture?

Interrupt priorities would be a thing the eirq has it's own vectors and wkpu has it's own interrupt vector.

gmarull commented:
> > WKPU is the "Wakeup Unit", which, according to the RM, it seems to be a unit responsible for configuring multiple wake-up sources.
> 
> Can you stop being so naive, and read further than the word wake-up. Indeed the functionality of the WKPU "Wakeup unit" is to provide wake-up events **and interrupts**. Which has been repeatedly commented here above.

I'm sorry, but it is your duty to provide all the details/explanations when there is a disagreement for whatever reason. Reviews take time, this is an open-source project, not some sort of vendor SDK. I'm happy to listen to all the details, but to me, some questions have not been answered.

gmarull commented:
> We've dismissed that suggestion from you before. Since it would be hiding the logic and makes it not transparent at all. Thus we make it explicit that the dts has to specifiy which interrupt controller has to be used for a certain pin.

So, if you already specify the interrupt controller, why do you need a special GPIO flag?

PetervdPerk-NXP commented:
> I'm sorry, but it is your duty to provide all the details/explanations when there is a disagreement for whatever reason. Reviews take time, this is an open-source project, not some sort of vendor SDK.

Which we are complying too, but if you're making it quite hard by speculating about "opening up possibilities" which aren't applicable here.  

> So, if you already specify the interrupt controller, why do you need a special GPIO flag?

The GPIO pin has to specify which interrupt controller the signal has to be routed to either EIRQ or WKPU that's on PIN basis, that's what the flag does.

gmarull commented:
> > I'm sorry, but it is your duty to provide all the details/explanations when there is a disagreement for whatever reason. Reviews take time, this is an open-source project, not some sort of vendor SDK.
> 
> Which we are complying too, but if you're making it quite hard by speculating about "opening up possibilities" which aren't applicable here.

can you please explain in detail **why** a user would want to do that (instead of going the normal route, assuming both are possible)?

>   To **route external interrupts to the WKPU controller**, the GPIO specifier must
  be supplied with the flag `NXP_S32_GPIO_INT_WKPU`. For example, the following
  snippet of DT source instructs the GPIO controller to route the interrupt on
  pin 9 of `gpioa` to the WKPU controller:
    #include <zephyr/dt-bindings/gpio/nxp-s32-gpio.h>
    &device {
        gpios = <&gpioa 9 (NXP_S32_GPIO_INT_WKPU | GPIO_ACTIVE_HIGH)>;
    };

PetervdPerk-NXP commented:
> can you please explain in detail why a user would want to do that (instead of going the normal route, assuming both are possible)?

Interrupt priorities would be a thing the eirq has it's own vectors and wkpu has it's own interrupt vector. So making it explicit which interrupt controller it's routed to makes it also clear which vector gets invoked.

gmarull commented:
> > can you please explain in detail why a user would want to do that (instead of going the normal route, assuming both are possible)?
> 
> Interrupt priorities would be a thing the eirq has it's own vectors and wkpu has it's own interrupt vector. So making it explicit which interrupt controller it's routed to makes it also clear which vector gets invoked.

Yes, got that part, but to do what? A practical example would be nice.

PetervdPerk-NXP commented:
> Yes, got that part, but to do what? A practical example would be nice.

So for example the buttons on the MR-CANHUBK344 can be both EIRQ or WKPU. Having a different irq priority for buttons would makes sense since it's less important then let's say a data ready interrupt from a sensor.

IRQ are grouped per 32 sources in WKPU, EIRQ etc, so you based on your circumstances you would want to map a pin to be WKPU or EIRQ.

fabiobaltieri commented:
> Interrupt priorities would be a thing the eirq has it's own vectors and wkpu has it's own interrupt vector. So making it explicit which interrupt controller it's routed to makes it also clear which vector gets invoked.

I don't really understand what that means. Let me rephrase the question, why can't the feature be designed to pick the right unit to manage the interrupt automatically as suggested? I also went through the change again and I don't think it's very clear why one would want to even do this in the first place, the documentation I see is all like "this can be done" but does not explain "why would one do it". Overall if it's possible to have the interrupt routing selected automatically without a device specific handling it'd definitely be a good thing, but even as it is the situation is not clear.

fabiobaltieri commented:
> I think this should be documented clearly, so users know why this exists. Also make it clear this should not be confused with wake-up capabilities of a pin.

@gmarull you think the implementation is ok as long as it gets documented clearly? To me that sounds like a fair use of GPIO flags.

@PetervdPerk-NXP so is this only for interrupt priority or there's other constraints/features associated with where the interrupt is routed?

dleach02 commented:
@gmarull, you are correct, it appears I need to escalate to the TSC the request to remove your block which I will unfortunately initiate.

But let me highlight that your initial block on the PR appeared to state that:

> In general I think we need a solution that is cross-vendor compatible. We have now STM32 using HAL directly in their power off sample, Nordic using gpio_interrupt_configure (it works by chance, because what is configured turns out to allow wake-up from pins) and now NXP introducing a vendor specific option.
>
> Please, let's propose something that can be made portable. There are enough SoCs with a similar feature.

This is contrary to our project's guidelines on reviewer expectations to avoid increasing the scope of a PR (https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html#reviewer-expectations). 

Instead of suggesting that the contributor of a PR solve a problem you have just identified, you should have posted an RFC to initiate discussion with the community. The community could then discuss and determine if the issue identified is truly a problem that needs a solution instead of independently deciding it is a problem and applying it to an arbitrary PR. 

This is a definition of increasing scope on a PR. 

Now that I guess it is clear this is not a GPIO/wakeup PR, your block is still in place and the collective folks engaging do not have a clear understanding on what is needed to clear your block.  Even @fabiobaltieri final comment could be interpreted as a non-blocking request that the authors can decided whether to implement or not (per our guidelines).

fabiobaltieri commented:
> Even @fabiobaltieri final comment could be interpreted as a non-blocking request that the authors can decided whether to implement or not (per our guidelines).

Ok but based on my current understanding I think that there's indeed a documentation/scoping gap that should be addressed. It's unfortunate we burned everyone's time on the whole wakeup naming discussion but I think that's also a consequence of the unclear scope for this feature.

The conversation is moving to something actionable, maybe let's keep it going in that direction.

nashif commented:
> @gmarull, you are correct, it appears I need to escalate to the TSC the request to remove your block which I will unfortunately initiate.

This is not correct. maintainer has the right to dismiss stale and irrelevant reviews without having to go to the TSC. We only go to the TSC if some escalates,either reviewer, maintainer or submitter. 

The notion that each dismissal of any review needs to go to the TSC will just bring the project to a halt.

dleach02 commented:
@fabiobaltieri I won't quibble this point and would concede to your view. My problem is that I don't want the authors to be making random changes trying to figure out what @gmarull wants or needs to get him to drop his block. The initial reason for his block was out of scope but now it is just unclear what he needs to get it cleared

fabiobaltieri commented:
> The initial reason for his block was out of scope but now it is just unclear what he needs to get it cleared

I think it's clear now? See https://github.com/zephyrproject-rtos/zephyr/pull/62623#issuecomment-1750720455, we'll get there. :-)

dleach02 commented:
> > The initial reason for his block was out of scope but now it is just unclear what he needs to get it cleared
> 
> I think it's clear now? See [#62623 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/62623#issuecomment-1750720455), we'll get there. :-) I have a couple comments myself on the code as well but I'm waiting for that to be clarified to (hopefully) avoid generating more noise.

This is you asking @gmarull if this would clear things up for him. Not necessarily the same as @gmarull clearly articulating what his block is and what is needed to clear it. I'll await his response. Thankyou for engaging on the PR.

gmarull commented:
> @gmarull, you are correct, it appears I need to escalate to the TSC the request to remove your block which I will unfortunately initiate.
> 
> But let me highlight that your initial block on the PR appeared to state that:
> 
> > In general I think we need a solution that is cross-vendor compatible. We have now STM32 using HAL directly in their power off sample, Nordic using gpio_interrupt_configure (it works by chance, because what is configured turns out to allow wake-up from pins) and now NXP introducing a vendor specific option.
> > Please, let's propose something that can be made portable. There are enough SoCs with a similar feature.
> 
> This is contrary to our project's guidelines on reviewer expectations to avoid increasing the scope of a PR (https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html#reviewer-expectations).
> 

I’ve blocked other PRs for similar reasons, and I’ll continue to do so if I think there’s a solid reason behind. We have the input subsystem nowadays because of excellent @fabiobaltieri work and in part because I blocked the introduction of partial/incomplete solutions before. 

I think this PR just failed to answer concerns properly from the beginning, though. And dismissals like this are definitely not acceptable IMHO. If you need this deployed quickly just maintain a fork, like some Zephyr vendors do or like many Linux BSPs, but please respect upstream reviewers work.

> Instead of suggesting that the contributor of a PR solve a problem you have just identified, you should have posted an RFC to initiate discussion with the community. The community could then discuss and determine if the issue identified is truly a problem that needs a solution instead of independently deciding it is a problem and applying it to an arbitrary PR.
> 
> This is a definition of increasing scope on a PR.

I understand vendors want to just push drivers and sometimes even impose their timelines upstream, but that’s not really my problem. If I believe something deserves a better explanation or implementation I have the right to block and to not be dismissed.

> 
> Now that I guess it is clear this is not a GPIO/wakeup PR, your block is still in place and the collective folks engaging do not have a clear understanding on what is needed to clear your block. Even @fabiobaltieri final comment could be interpreted as a non-blocking request that the authors can decided whether to implement or not (per our guidelines).

Whether it’s used for wake-up or not, current code allows to configure a GPIO as a wake-up source. It looks like you offered convincing explanations that this also has some other utilities, like tweaking IRQ priorities (unfortunately it took too much time to get that information - not my fault). So it’s a matter these are documented clearly, and, that it is explicitly mentioned that these flags shall not be used to configure GPIOs as wake-up sources (I’m definitely going to block that if this ever comes for the explained reasons).

fabiobaltieri commented:
> This is you asking @gmarull if this would clear things up for him. Not necessarily the same as @gmarull clearly articulating what his block is and what is needed to clear it.

I'm trying to think one step forward and see if there could be anything apart from a documentation gap explaining how to use that feature once the scope has been defined with ambiguity (which is the other question I asked there). The documentation gap I think has to be addressed regardless, I'm looking at the diff from a user perspective and it's unclear to me why I would use the feature.

fabiobaltieri commented on include/zephyr/dt-bindings/gpio/nxp-s32-gpio.h:
```diff
@@ -0,0 +1,41 @@
+ * @{
+ */
+
+/** Interrupt routed to the SIUL2 External Interrupt controller */
+#define NXP_S32_GPIO_INT_EIRQ	(0x0U << NXP_S32_GPIO_INT_TYPE_POS)
```
Hey I don't quite see a point in defining this. It's not used in the code anywhere, defining it opens the possibility of one setting both this and the `WKPU` flag which makes no sense and should be handled. I think it should just be dropped, just define`NXP_S32_GPIO_INT_WKPU` and clarify in the documentation that interrupt are routed by default through the EIRQ/SIUL2 and go to WKPU if the flag is specified.

While I'm at it: the whole EIRQ/SIUL2 naming is a fairly confusing, may not be exposed too much if you drop this one (it's odd to have SIUL2 in the documentation and EIRQ in the flag name).

manuargue replied:
it makes sense, dropped flag as requested

fabiobaltieri commented on dts/bindings/gpio/nxp,s32-gpio.yaml:
```diff
@@ -1,7 +1,23 @@
+  be supplied with the flag `NXP_S32_GPIO_INT_WKPU`. For example, the following
+  snippet of DT source instructs the GPIO controller to route the interrupt on
+  pin 9 of `gpioa` to the WKPU controller:
+
```
this is probably a good spot to expand on why a user would want to do this and I guess make explicit that there's no wake-up functionalities (for now)

manuargue replied:
added details as requested

fabiobaltieri replied:
Thanks!

dleach02 commented:
@gmarull, 

It is not a good representation for the project to air this out in a PR. We will take this to a more appropriate forum to discuss and ultimately to clarify.

PetervdPerk-NXP commented:
> so is this only for interrupt priority or there's other constraints/features associated with where the interrupt is routed?

Only the interrupt priority of the affected vector where the pin interrupt will be routed to.

fabiobaltieri commented:
> > so is this only for interrupt priority or there's other constraints/features associated with where the interrupt is routed?
> 
> Only the interrupt priority of the affected vector where the pin interrupt will be routed to.

Ok, I think that if you adjust the documentation so that
- it is clear that that this does not support wake up functionalities ATM
- the use case is clearly explained (prob in the binding doc)

then it should be good to go. See my comment on the flag as well.

gmarull commented:
> > Only the interrupt priority of the affected vector where the pin interrupt will be routed to.
> 
> Ok, I think that if you adjust the embedded documentation so that
> 
> * it is clear that that this does not support wake up functionalities ATM

Does not support **and it is not meant to do so**, so we don't need to have the same discussion again. I want to avoid another vendor-specific wake-up mechanism, we have enough cases in-tree to find a common ground and so do better than random vendor solutions.

decsny commented:
@gmarull It seems like you have a good vision here, and it is unfortunate that the development of the project comes to a halt until someone implements that vision.

gmarull commented:
> @gmarull It seems like you have a good vision here, and it is unfortunate that the development of the project comes to a halt until someone implements that vision.

It looks like you didn't want to implement wake-up in the end, or did you? I'm now confused. If not, I'm just asking for a couple of lines in the documentation, that's it.

decsny commented:
> It looks like you didn't want to implement wake-up in the end, or did you? I'm now confused. If not, I'm just asking for a couple of lines in the documentation, that's it.

You have me confused for someone else, I never commented in this thread before.

gmarull commented:
> > It looks like you didn't want to implement wake-up in the end, or did you? I'm now confused. If not, I'm just asking for a couple of lines in the documentation, that's it.
> 
> You have me confused for someone else, I never commented in this thread before.

Guys, just update the PR with some minimal docs as requested, PR can be unblocked, and we can all move on with our lives. Not a big deal. Please let me know when this has been done, so I can re-review and unblock if all looks reasonable.

fabiobaltieri commented:
Folks I think at this point the requests are clear and reasonable (but please reply back if that's not the case and let's try to work on it). I think it's better use of everyone time to focus on the get this PR moving. It's unfortunate that it took some friction to get there but there's no need to create more.

decsny commented:
> Folks I think at this point the requests are clear and reasonable (but please reply back if that's not the case and let's try to work on it). I think it's better use of everyone's time to focus on the get this PR moving. It's unfortunate that it took some friction to get there but there's no need to create more.

I agree, thank you to @gmarull for the documentation suggestion

manuargue commented:
> > Only the interrupt priority of the affected vector where the pin interrupt will be routed to.
> 
> Ok, I think that if you adjust the embedded documentation so that
> 
> * it is clear that that this does not support wake up functionalities ATM
> * the use case is clearly explained (prob in the binding doc), also make it clear what setting has the highest priority then, not obvious now
> 
> then it should be good to go. See my comment on the flag as well.

Thanks for the review @fabiobaltieri 

I've pushed documentation updates to address these issues. Please help me have a look and let me know if further clarification/explanations are needed. I've tried to cover everything that was mentioned on this PR comments.

fabiobaltieri reviewed (dismissed):
Thanks for your patience @manuargue.

gmarull commented on dts/bindings/gpio/nxp,s32-gpio.yaml:
```diff
@@ -1,7 +1,39 @@
+
+  The `NXP_S32_GPIO_INT_WKPU` flag is intended exclusively for specifying WKPU
+  as the interrupt controller for the corresponding GPIO. It's worth noting that
+  despite being named WKPU, this driver does not provide support for wake-up
+  events.
```
```suggestion
  despite being named WKPU, the flag is not meant to configure GPIOs as wake-up sources.
```

nit: bindings are dt schemas, so not strictly tied to drivers.

manuargue replied:
noted, suggested patch applied

jhedberg commented:
@dleach02 @manuargue is this something that you're hoping would go in for 3.5? If so, please set the target milestone.

jhedberg commented:
Why does this still have the DNM label?

fabiobaltieri commented:
TSC passed this for merge during stabilization. Thanks everyone for your patience.
